### ISSUE TITLE
Multiple entity_save() calls and spinning loader icon

### ISSUE THREAD

**Issue by chrisschaub** [signalpoint/DrupalGap#424]
I've noticed that the file entity_save does not trigger the loader -- or maybe it happens and gets killed. Any suggestions on where to begin? Just need to show saving message for each file saved. I've tried placing various places, but it keeps getting squashed.

**Comment by chrisschaub** [comment]
It's like entity preprocess doesn't always fire.

**Comment by signalpoint** [comment]
@chrisschaub Is this related to uploading multiple files at once? If so, I think the async nature of it starts the loading spinner, but soon as the first one finishes, it probably hides the spinner, and then by that point all the others may have already started, so the spinner was already loaded, and then it gets hidden by the first completion.

I'm not sure though, I haven't had a chance to test out the multiple file upload stuff yet.

**Comment by chrisschaub** [comment]
No, it happens either way. And there are other async things happening with
fetching entities that could cuase this too. It does seem like one entity
postprocess could stop other entity pre-process calls while they're
happening. It would be really slick to queue up an entity process global
object, use whenall and process them, then do final stuff when all the
defers come back. That way you'd get a true stop/start without chaining or
async:false.

On Wed, Nov 26, 2014 at 10:25 AM, Tyler Frankenstein <
notifications@github.com> wrote:

> @chrisschaub https://github.com/chrisschaub Is this related to
> uploading multiple files at once? If so, I think the async nature of it
> starts the loading spinner, but soon as the first one finishes, it probably
> hides the spinner, and then by that point all the others may have already
> started, so the spinner was already loaded, and then it gets hidden by the
> first completion.
> 
> I'm not sure though, I haven't had a chance to test out the multiple file
> upload stuff yet.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/signalpoint/DrupalGap/issues/424#issuecomment-64661374
> .

## 

Christopher Schaub
http://chris.schaub.com

**Comment by pierpa68** [comment]
@signalpoint This issue still seems open... and in fact I found this problem saving nodes with images. It is true even if there is just one image in the entity.

Any guess?

**Comment by signalpoint** [comment]
@pierpa68 do you have a code sample for the situation? Last I checked a simple single value image field on nodes worked, I don't believe I've ever tried it on a different entity type (I think the user entity type works fine too for file fields). There is a long standing (_sorry Chris_) pull request from @chrisschaub here for handling multiple images: https://github.com/signalpoint/DrupalGap/pull/406

I could see multiple entity_save() calls working if the iterator were in the success callback. Things could get a little squirrel-ey if just using a for/while/each to make the calls, but can also work.

**Comment by pierpa68** [comment]
@signalpoint My test is on simple one-image node saving... no coding involved.

**Comment by pierpa68** [comment]
@signalpoint I spent some time today debugging this.
It seems a problem with 

```
setTimeout(function() {
      console.log("loading show");
      $.mobile.loading('show', options);
      drupalgap.loading = true;     
    }, 1);
```

as the function containing it ( `drupalgap_loading_message_show()`) is called BEFORE saving the file, but the function `$.mobile.loading('show', options)` itself is fired AFTER the the file is saved.

In fact, setting `async: true` in `function _image_field_form_process` makes the spinning loader appear.

I've tried several solutions, but nothing seems satisfactory. 
The simpler is to put the `file_save` inside a `setTimeout` function with some delay, but doing so the spinner appears but "freezed".

What about using some other plugin, such as the BlockUI one?

**Comment by pierpa68** [comment]
@signalpoint Solved temporary using `cordova-plugin-spinnerdialog` which has the advantage to look native on mobile application and to avoid user interaction, so can re-enable async file transfer.

In order to keep maximum compatibility even with web application, I extended $.mobile in my custom module (inside `hook_deviceredy` ):

```
if (SpinnerDialog) {
        $.extend($.mobile, {
            loading: function() {
                // Show/hide spinner
                var arg = arguments ? arguments[0] : '';
                var text = arguments ? (arguments[1] ? arguments[1].text : '' ) : '';

                if (arg == 'show') SpinnerDialog.show("",text,false);
                else if (arg == 'hide') SpinnerDialog.hide();           

                 return { loader: function() { } }
            }
        }); 
    }
```

In drupalgap.js (src/modules/image) I just changed function `_image_field_form_process` setting `async: true` on the call to `file_save` function.

HTH

**Comment by pierpa68** [comment]
Note: Last parameter of `SpinnerDialog.show` should be set to `true` in order to lock user interface.

Sorry...

**Comment by signalpoint** [comment]
@pierpa68 Interesting, thank you for sharing. Would you consider creating e.g. the "spinnerdialog" module for DrupalGap? Then you could put the extension on $.mobile into that module's device ready.

After that, you could use the `hook_services_preproces()` (http://api.drupalgap.org/7/global.html#hook_services_preprocess) in spinnerdialog.js to alter the service call, and set `async` to true.

This would be an awesome contribution. What do you think?
